llms/mlx_lm/tuner/trainer.py

```diff
@@ -0,0 +1,224 @@
+
+    def train(self):
+        # Create value and grad function for loss
+        loss_value_and_grad = nn.value_and_grad(self.model, self.loss)
+
+        losses = []
+        n_tokens = 0
+        print("Starting training..., iters:", self.args.iters)
+        # Main training loop
+        start = time.perf_counter()
+        for it, batch in zip(
+            range(self.args.iters),
+            iterate_batches(
+                self.train_dataset,
+                self.tokenizer,
+                self.args.batch_size,
+                self.args.max_seq_length,
+            ),
+        ):
+            # Forward and backward pass
+            (lvalue, toks), grad = loss_value_and_grad(self.model, *batch)
+
+            # Model update
+            self.optimizer.update(self.model, grad)
+
+            mx.eval(self.model.parameters(), self.optimizer.state, lvalue)
+
+            # Record loss
+            losses.append(lvalue.item())
+            n_tokens += toks.item()
+
+            # Report training loss if needed
+            if (it + 1) % self.args.steps_per_report == 0:
+                train_loss = np.mean(losses)
+
+                stop = time.perf_counter()
+                print(
+                    f"Iter {it + 1}: Train loss {train_loss:.3f}, "
+                    f"It/sec {self.args.steps_per_report / (stop - start):.3f}, "
+                    f"Tokens/sec {float(n_tokens) / (stop - start):.3f}"
+                )
+                losses = []
+                n_tokens = 0
+                start = time.perf_counter()
+
+            # Report validation loss if needed
+            if it == 0 or (it + 1) % self.args.steps_per_eval == 0:
+                stop = time.perf_counter()
+                val_loss = evaluate(
+                    self.model,
+                    self.val_dataset,
+                    self.loss,
+                    self.tokenizer,
+                    self.args.batch_size,
+                    self.args.val_batches,
+                )
+                print(
+                    f"Iter {it + 1}: "
+                    f"Val loss {val_loss:.3f}, "
+                    f"Val took {(time.perf_counter() - stop):.3f}s"
+                )
+
+                start = time.perf_counter()
+
+                # Save adapter weights if needed
+                if (it + 1) % self.args.steps_per_save == 0:
+                    self.save_adapter()
+                print(
+                    f"Iter {it + 1}: Saved adapter weights to {os.path.join(self.save_dir, self.args.adapter_file)}."
+                )
+
+    def save_adapter(
+        self,
+    ):
+        if not self.save_dir.is_dir():
+            raise ValueError(f"The provided path {self.save_dir} is not a directory.")
+
+        flattened_tree = tree_flatten(self.model.trainable_parameters())
+
+        mx.savez(self.save_dir / self.adapter_file, **dict(flattened_tree))
+        # Save the LoRA layer names
+        param_names = list(
```

Comment:
Save the additional Lora layer parameter name in the JSON file, along with the adapter weight. So, if we can find a way to apply the Lora linear according to the JSON later, that way, in fuse.py or when generating text from model with an adapter, we won't have to manually apply the Lora linear. We should be able to read from JSON and attach Lora to the base model. This will make distributing Lora and inference easier.